OpenRA.Game/Widgets/Widget.cs

```diff
@@ -110,20 +110,54 @@ public static bool HandleInput(MouseInput mi)
 
 		public static bool HandleKeyPress(KeyInput e)
 		{
-			if (KeyboardFocusWidget != null)
-				return KeyboardFocusWidget.HandleKeyPressOuter(e);
+			if (KeyboardFocusWidget.Count > 0 && KeyboardFocusWidget.Peek() != null)
+				return KeyboardFocusWidget.Peek().HandleKeyPressOuter(e);
 
 			return Root.HandleKeyPressOuter(e);
 		}
 
 		public static bool HandleTextInput(string text)
 		{
-			if (KeyboardFocusWidget != null)
-				return KeyboardFocusWidget.HandleTextInputOuter(text);
+			if (KeyboardFocusWidget.Count > 0 && KeyboardFocusWidget.Peek() != null)
+				return KeyboardFocusWidget.Peek().HandleTextInputOuter(text);
 
 			return Root.HandleTextInputOuter(text);
 		}
 
+		public static bool CheckKeyboardFocus(Widget w)
+		{
+			if (KeyboardFocusWidget.Count == 0)
+				return false;
+
+			if (KeyboardFocusWidget.Peek() == w)
+				return true;
+
+			return false;
```

Comment:
If `Peek() == w -> true` is correct, then you can just `return KeyboardFocusWidget.Peek() == w`.